sbi/utils/sbiutils.py

```diff
@@ -171,24 +173,23 @@ def sample_posterior_within_prior(
     num_remaining, num_sampled_total = num_samples, 0
     accepted, acceptance_rate = [], float("Nan")
     leakage_warning_raised = False
-    # In each iteration of the loop we sample the remaining number of samples from the
-    # posterior. Some of these samples have 0 probability under the prior, i.e. there
-    # is leakage (acceptance rate<1) so sample again until reaching `num_samples`.
-    while num_remaining > 0:
 
-        candidates = posterior_nn.sample(num_remaining, context=x)
-        # TODO we need this reshape here because posterior_nn.sample sometimes return
-        # leading singleton dimension instead of (num_samples), e.g., (1, 10000, 4)
-        # instead of (10000, 4). This can't be handled by MultipleIndependent, see #141.
-        candidates = candidates.reshape(num_remaining, -1)
-        num_sampled_total += num_remaining
+    # Sample once to cover cases without leakage.
+    samples = sample_and_reject(posterior_nn, x, prior, num_samples=num_samples)
```

Comment:
should be `max(num_samples, sampling_batch_size)` to allow for many samples that would otherwise fill RAM

Reply:
Good catch, so I guess `sample_and_reject` would need to take a batch size argument in that case

Reply:
will push a different solution.